### PR TITLE
[TF:TRT] Enable Dynamic Shape Support for ConvertResize

### PR DESCRIPTION
**Pull request by wraveane**
The Resize operation now supports full dynamic shapes. This is done by slicing the input tensor's shape, replacing the H and W dims with the corresponding Resize op scales, and concatenating the dims back to a shape tensor that can then be given as a second input to the TRT Resize layer.

In case of static shapes, the previous method remains, where the output dims are simply given as a parameter to the TRT Resize layer instead.

A new use case is now also supported, where the "size" may now be given as a tensor input instead of constant weights. This would often happen when working with dynamic shapes, as the output size may need to be calculated on the fly through shape ops. E.g., resizing a tensor by a scale factor of 2 would require the size to be calculated dynamically in the graph based on the input tensor's shape, thus not a constant input that is known during building.

The corresponding test was also updated to now expect all 3 TRT modes to pass: kImplicitBatch, kExplicitBatch and kDynamicShape.

Tracker: #45481

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6674]: Let's merge this if-stmt with the previous if-stmt, by moving the code between the two if-stmts to before the first if-stmt.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6615]: This comment should go inside the if branch below

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6618]: This line should go inside the else branch below.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6617]: This line should go inside the if branch below.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6668]: Let's revise this comment:
If the input shape is known at build time, pass the static output shape to the resize layer as expected output dimensions.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6671]: Let's revise the comment:
Otherwise, pass the output shape tensor to the resize layer as an input.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6619]: Shall we remove params->use_implicit_batch here?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6674]: Actually, it's not possible to merge them, in case of dynamic shape, the concatenation of the shape tensor needs to be done before the resize layer itself, but then setting the output dimensions and/or adding the secondary input can only be done after the resize layer is created.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6615]: The comment applies to both branches, as it's a similar logic in both cases: H_out and W_out are overridden by the scale attribute in both cases.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6619]: This branch should also run when running in implicit batch mode, as in that case the output dimensions need to be explicitly set as well. However if you think it's cleaner -- I could change the conditionals to be more "descriptive" by checking for whether `output_shape_dims` or `output_shape_tensor` have been set in the if statement above, and run the corresponding branch in either case.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6618]: Both output_shape_dims and output_shape_tensor need to be on this outer scope because these objects are used on the other if statement below.

**Comment by wraveane** [comment]
@googlebot I signed it!

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6619]: I have two questions:
(1) the block of code inside this if-branch also works and executes when (use_implicit_batch==false && HasStaticShape(inputs_dims)) before your PR, see [code](https://github.com/tensorflow/tensorflow/blob/2825dbfe950cd514828da4bfed81a2402d779b35/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L6609-L6612). Am I right here? If not, what are the reason that this code don't work for use_implicit_batch=false? I couldn't see that...
(2) This is not introduced by your PR, but when I read [this code](https://github.com/tensorflow/tensorflow/blob/2825dbfe950cd514828da4bfed81a2402d779b35/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L6618-L6619), I think it has a problem for use_implicit_batch=true as we remove the batch dimension and should only have three dimensions in the input tensor.  Do you agree?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: I see what you meant. Can you (1) define "nvinfer1::IResizeLayer* layer = nullptr" before the if-stmt, and let the branches create the layer,  and (2) move the relevant code from [this block](https://github.com/tensorflow/tensorflow/blob/2825dbfe950cd514828da4bfed81a2402d779b35/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L6635-L6641) to after the if-stmt?

Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: I think I understand what you're asking, but I'm not sure this would result in cleaner code (unless I'm overlooking something obvious). The resize layer would still have to be created the same way on both branches -- basically because the `addResize()` method only takes a single input tensor anyway, so the second input would still need to be provided through a separate statement regardless. Please correct me if I missed the point of what you asked of course.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6619]: Sure, so:
(1) When running in implicit batch mode, all non-batch dimensions are known at build time. So that is enough to be able to calculate the output shape dimensions. So the implicit batch check there acts as a flag that allows this dimension calculation to happen at build time. In any case, this logic is slightly modified in the commit I'm pushing soon for dynamic output size support.
(2) Kind of -- So in implicit batch mode, during build time, the batch dimension is skipped and all build-time calculations must happen ignoring that dimension. However at run-time, the batch dimension does exist obviously, so any runtime computations must take the N batch dimension into account. So the transpose layer that you pointed out must be configured with how the layer will operate at runtime, and so the transpose dims must be configured taking the batch dimension in account.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6575]: This code seems to imply is that if the rank is not 2, it is 0 ("shape=[]" in the error message). The error message it NOT added by this PR, so this is not your fault. Can we change the error message to "Resize requires 2D values for the new size"?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6587]: Similar to the error message in the if-branch.

**Comment by bixia1** [comment]
I couldn't figure out a way to reply to https://github.com/tensorflow/tensorflow/pull/51462#discussion_r693990638, so I am putting my response here. I don't feel you have answered my two questions, we need to schedule a GCV to discuss this.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6565]: NIT: we can replace the use of static_input_shape && const_output_size with
 const bool static_output_shape = HasStaticShape(inputs_dims) && const_output_size

the put this new variable closer to where it is used.

**Review by bixia1 — COMMENTED** [review]
Would you please squash all the commits into one? This is to help simplify the auto generated message that goes with the commit, and it is usually done right before the PR is approved.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6620]: I was thinking this comment was talking about calculating output_shape_dims, the else branch actually calculates a output shape tensor. Let's do this:
s/Calculate output dimensions./Calculate the output shape as static dimensions or a shape tensor:/

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6621]: this comment assumes we merge the two if-stmt

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6622]: Same, the comment assume we merge the two if-stmts.

**Comment by bixia1** [comment]
Looks good to me. Thanks for working on this!
When you address the last minor comment, please also squash the commits into one as request [here](https://github.com/tensorflow/tensorflow/pull/51462#pullrequestreview-736649531).

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Ah right, I guess keeping the word "dimensions" could be mis-interpreted given the two different methods there are now. I updated this.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: This error message was updated in a commit yesterday.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: This error message was updated in a commit yesterday.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: We discussed this offline, I think the current version now addresses your concern with how implicit batch is checked for.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Please let me know if there was anything else to do for this comment, as I think there's not much else that can be done with the way the layer is created in the current version.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (wraveane) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Yes, this might be a bit cleaner indeed. Updated it as per your suggestion.

**Comment by wraveane** [comment]
> Looks good to me. Thanks for working on this!
> When you address the last minor comment, please also squash the commits into one as request [here](https://github.com/tensorflow/tensorflow/pull/51462#pullrequestreview-736649531).

@bixia1 : Ok great. I think I updated everything as per all the comments. Please let me know if anything else is still needed.

As for squashing, because I had to merge master into my branch to handle a conflict, squashing the commits now would require me to do a force push of the branch I think. Is that ok? Otherwise, I think github always allows the option to squash before merging once the review is complete.

**Comment by bixia1** [comment]
Is this something I have a way to indicate on github "I think github always allows the option to squash before merging once the review is complete" ? I don't actually do anything to merge, but let the "automatic process to do the job".

If I can't  figure out how to indicate "squash before merging", I need you to squash your commits. I have reviewed your new changes, you are free to do the merge and squash.

**Comment by wraveane** [comment]
Ok, all good now, commits have been squashed.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**
